<a href="https://colab.research.google.com/github/MoeSaleh24/School-Projects/blob/main/Diamond_Prediction_(LR%2CKNN%2CDT%2CRF%2CBT%2CXGB).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from scipy import stats
import math


path = "/content/drive/MyDrive/INNFO367/Week 8/diamonds_train.csv"
df = pd.read_csv(path)

df.head()

shape        cut color clarity  weight_ct     fluorescence  \
0  Emerald  Very Good     D     SI1       1.70  Medium BlueInfo   
1    Round      Ideal     F      IF       1.07        FaintInfo   
2  Radiant  Very Good     G    VVS1       1.05         NoneInfo   
3    Round      Ideal     E    VVS2       2.85         NoneInfo   
4    Round      Ideal     D     SI1       0.35         NoneInfo   

   length_width_ratio     polish   symmetry                    girdle cutlet  \
0                1.29  Very Good  Very Good        Thin to Very Thick   None   
1                1.00  Excellent  Excellent    Thin to Slightly Thick   None   
2                1.20  Excellent  Excellent   Slightly Thick to Thick   None   
3                1.01  Excellent  Excellent  Medium to Slightly Thick   None   
4                1.00  Excellent  Excellent  Medium to Slightly Thick   None   

   price  depth  table  length  width  height  
0  13519   68.1   70.0    7.94   6.17    4.20  
1   8842   62.4   56.0    6.52   6.54    4.07  
2   4817   71.3   68.0    6.15   5.12    3.65  
3  63161   62.7   58.0    9.06   9.00    5.66  
4    653   59.9   59.0    4.57   4.59    2.74

In [ ]:
df = df.sample(n=30000, random_state= 2)

df.head()

shape        cut color clarity  weight_ct fluorescence  \
34535   Princess  Very Good     I     VS1       1.70     NoneInfo   
128812      Pear  Very Good     E    VVS1       1.75     NoneInfo   
89439      Round      Ideal     F    VVS1       1.06     NoneInfo   
88549      Round      Ideal     H     SI1       0.50     NoneInfo   
127858     Round      Ideal     F     VS1       1.00     NoneInfo   

        length_width_ratio     polish   symmetry                    girdle  \
34535                 1.01  Excellent  Very Good       Very Thin to Medium   
128812                1.59  Excellent  Very Good        Thin to Very Thick   
89439                 1.01  Excellent  Excellent  Medium to Slightly Thick   
88549                 1.01  Excellent  Excellent  Medium to Slightly Thick   
127858                1.01  Excellent  Excellent    Thin to Slightly Thick   

       cutlet  price  depth  table  length  width  height  
34535    None   9733   72.5   72.0    6.40   6.36    4.61  
128812   None  22920   60.7   57.0    0.78   6.77    4.11  
89439    None   9612   59.5   59.0    6.69   6.64    3.97  
88549    None   1345   62.2   58.0    5.07   5.10    3.16  
127858   None   7964   63.0   56.0    6.32   6.37    3.99

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30000 entries, 34535 to 77618
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   shape               30000 non-null  object 
 1   cut                 30000 non-null  object 
 2   color               30000 non-null  object 
 3   clarity             30000 non-null  object 
 4   weight_ct           30000 non-null  float64
 5   fluorescence        30000 non-null  object 
 6   length_width_ratio  30000 non-null  float64
 7   polish              30000 non-null  object 
 8   symmetry            30000 non-null  object 
 9   girdle              30000 non-null  object 
 10  cutlet              30000 non-null  object 
 11  price               30000 non-null  int64  
 12  depth               30000 non-null  float64
 13  table               30000 non-null  float64
 14  length              30000 non-null  float64
 15  width               30000 non-null  float64
 16  

In [ ]:
df = pd.get_dummies(df, drop_first=True)

df.head()

weight_ct  length_width_ratio  price  depth  table  length  width  \
34535        1.70                1.01   9733   72.5   72.0    6.40   6.36   
128812       1.75                1.59  22920   60.7   57.0    0.78   6.77   
89439        1.06                1.01   9612   59.5   59.0    6.69   6.64   
88549        0.50                1.01   1345   62.2   58.0    5.07   5.10   
127858       1.00                1.01   7964   63.0   56.0    6.32   6.37   

        height  shape_Cushion  shape_Emerald  ...  girdle_Very Thin to Thin  \
34535     4.61              0              0  ...                         0   
128812    4.11              0              0  ...                         0   
89439     3.97              0              0  ...                         0   
88549     3.16              0              0  ...                         0   
127858    3.99              0              0  ...                         0   

        girdle_Very Thin to Very Thick  \
34535                                0   
128812                               0   
89439                                0   
88549                                0   
127858                               0   

        girdle_Very Thin to Very Thick Faceted  girdle_Very Thin to Very Thin  \
34535                                        0                              0   
128812                                       0                              0   
89439                                        0                              0   
88549                                        0                              0   
127858                                       0                              0   

        cutlet_Medium  cutlet_None  cutlet_Pointed  cutlet_Slightly Large  \
34535               0            1               0                      0   
128812              0            1               0                      0   
89439               0            1               0                      0   
88549               0            1               0                      0   
127858              0            1               0                      0   

        cutlet_Small  cutlet_Very Small  
34535              0                  0  
128812             0                  0  
89439              0                  0  
88549              0                  0  
127858             0                  0  

[5 rows x 126 columns]

In [ ]:
X = df.drop('price', axis=1)     # Features (independent variables)
y = df['price']

X.head()

weight_ct  length_width_ratio  depth  table  length  width  height  \
34535        1.70                1.01   72.5   72.0    6.40   6.36    4.61   
128812       1.75                1.59   60.7   57.0    0.78   6.77    4.11   
89439        1.06                1.01   59.5   59.0    6.69   6.64    3.97   
88549        0.50                1.01   62.2   58.0    5.07   5.10    3.16   
127858       1.00                1.01   63.0   56.0    6.32   6.37    3.99   

        shape_Cushion  shape_Emerald  shape_Heart  ...  \
34535               0              0            0  ...   
128812              0              0            0  ...   
89439               0              0            0  ...   
88549               0              0            0  ...   
127858              0              0            0  ...   

        girdle_Very Thin to Thin  girdle_Very Thin to Very Thick  \
34535                          0                               0   
128812                         0                               0   
89439                          0                               0   
88549                          0                               0   
127858                         0                               0   

        girdle_Very Thin to Very Thick Faceted  girdle_Very Thin to Very Thin  \
34535                                        0                              0   
128812                                       0                              0   
89439                                        0                              0   
88549                                        0                              0   
127858                                       0                              0   

        cutlet_Medium  cutlet_None  cutlet_Pointed  cutlet_Slightly Large  \
34535               0            1               0                      0   
128812              0            1               0                      0   
89439               0            1               0                      0   
88549               0            1               0                      0   
127858              0            1               0                      0   

        cutlet_Small  cutlet_Very Small  
34535              0                  0  
128812             0                  0  
89439              0                  0  
88549              0                  0  
127858             0                  0  

[5 rows x 125 columns]

In [ ]:
train_X, valid_X, train_y, valid_y = train_test_split(X, y, test_size=0.3, random_state=1)

In [ ]:
re_prices = LinearRegression()

re_prices.fit(train_X, train_y)

LinearRegression()

In [ ]:
train_X.head()

weight_ct  length_width_ratio  depth  table  length  width  height  \
28919        0.50                1.00   63.3   58.0    4.98   4.96    3.15   
92484        0.59                1.38   67.8   58.0    6.34   4.58    3.11   
142209       1.00                1.01   63.4   57.0    6.31   6.38    4.02   
57388        1.00                1.05   68.9   60.0    5.67   5.41    3.73   
59796        0.40                1.01   63.3   58.0    4.67   4.70    2.97   

        shape_Cushion  shape_Emerald  shape_Heart  ...  \
28919               0              0            0  ...   
92484               0              0            0  ...   
142209              0              0            0  ...   
57388               1              0            0  ...   
59796               0              0            0  ...   

        girdle_Very Thin to Thin  girdle_Very Thin to Very Thick  \
28919                          0                               0   
92484                          0                               0   
142209                         0                               0   
57388                          0                               0   
59796                          0                               0   

        girdle_Very Thin to Very Thick Faceted  girdle_Very Thin to Very Thin  \
28919                                        0                              0   
92484                                        0                              0   
142209                                       0                              0   
57388                                        0                              0   
59796                                        0                              0   

        cutlet_Medium  cutlet_None  cutlet_Pointed  cutlet_Slightly Large  \
28919               0            1               0                      0   
92484               0            1               0                      0   
142209              0            1               0                      0   
57388               0            1               0                      0   
59796               0            1               0                      0   

        cutlet_Small  cutlet_Very Small  
28919              0                  0  
92484              0                  0  
142209             0                  0  
57388              0                  0  
59796              0                  0  

[5 rows x 125 columns]

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, max_error, r2_score, explained_variance_score

In [ ]:
pred_y = re_prices.predict(valid_X)

In [ ]:
print('Regression model performance evaluation')
print('R2: ', r2_score(valid_y, pred_y))
print('MAE: ', mean_absolute_error(valid_y, pred_y))
print('RMSE: ', math.sqrt(mean_squared_error(valid_y, pred_y)))
print('Max error: ', max_error(valid_y, pred_y))

Regression model performance evaluation
R2:  0.40077331320237364
MAE:  7887.698542832007
RMSE:  16255.798470405343
Max error:  597965.5843113931


In [ ]:
df = pd.get_dummies(df)

X.head()

weight_ct  length_width_ratio  depth  table  length  width  height  \
34535        1.70                1.01   72.5   72.0    6.40   6.36    4.61   
128812       1.75                1.59   60.7   57.0    0.78   6.77    4.11   
89439        1.06                1.01   59.5   59.0    6.69   6.64    3.97   
88549        0.50                1.01   62.2   58.0    5.07   5.10    3.16   
127858       1.00                1.01   63.0   56.0    6.32   6.37    3.99   

        shape_Cushion  shape_Emerald  shape_Heart  ...  \
34535               0              0            0  ...   
128812              0              0            0  ...   
89439               0              0            0  ...   
88549               0              0            0  ...   
127858              0              0            0  ...   

        girdle_Very Thin to Thin  girdle_Very Thin to Very Thick  \
34535                          0                               0   
128812                         0                               0   
89439                          0                               0   
88549                          0                               0   
127858                         0                               0   

        girdle_Very Thin to Very Thick Faceted  girdle_Very Thin to Very Thin  \
34535                                        0                              0   
128812                                       0                              0   
89439                                        0                              0   
88549                                        0                              0   
127858                                       0                              0   

        cutlet_Medium  cutlet_None  cutlet_Pointed  cutlet_Slightly Large  \
34535               0            1               0                      0   
128812              0            1               0                      0   
89439               0            1               0                      0   
88549               0            1               0                      0   
127858              0            1               0                      0   

        cutlet_Small  cutlet_Very Small  
34535              0                  0  
128812             0                  0  
89439              0                  0  
88549              0                  0  
127858             0                  0  

[5 rows x 125 columns]

In [ ]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
X_ = scaler.fit_transform(X)

X_rescaled = pd.DataFrame(X_, columns=X.columns)

In [ ]:
X_rescaled.describe()

weight_ct  length_width_ratio         depth         table  \
count  30000.000000        30000.000000  30000.000000  30000.000000   
mean       0.032164            0.151338      0.430348      0.276549   
std        0.041142            0.105901      0.129969      0.129139   
min        0.000000            0.000000      0.000000      0.000000   
25%        0.009101            0.102857      0.362069      0.200000   
50%        0.017666            0.108571      0.396552      0.228571   
75%        0.041756            0.114286      0.455172      0.314286   
max        1.000000            1.000000      1.000000      1.000000   

             length         width        height  shape_Cushion  shape_Emerald  \
count  30000.000000  30000.000000  30000.000000   30000.000000   30000.000000   
mean       0.562132      0.535791      0.345193       0.035267       0.046567   
std        0.144305      0.121004      0.080437       0.184456       0.210712   
min        0.000000      0.000000      0.000000       0.000000       0.000000   
25%        0.465465      0.448448      0.284687       0.000000       0.000000   
50%        0.529530      0.510511      0.322713       0.000000       0.000000   
75%        0.644645      0.614615      0.397739       0.000000       0.000000   
max        1.000000      1.000000      1.000000       1.000000       1.000000   

        shape_Heart  ...  girdle_Very Thin to Thin  \
count  30000.000000  ...              30000.000000   
mean       0.014700  ...                  0.002100   
std        0.120351  ...                  0.045778   
min        0.000000  ...                  0.000000   
25%        0.000000  ...                  0.000000   
50%        0.000000  ...                  0.000000   
75%        0.000000  ...                  0.000000   
max        1.000000  ...                  1.000000   

       girdle_Very Thin to Very Thick  girdle_Very Thin to Very Thick Faceted  \
count                    30000.000000                            30000.000000   
mean                         0.002600                                0.000067   
std                          0.050925                                0.008165   
min                          0.000000                                0.000000   
25%                          0.000000                                0.000000   
50%                          0.000000                                0.000000   
75%                          0.000000                                0.000000   
max                          1.000000                                1.000000   

       girdle_Very Thin to Very Thin  cutlet_Medium   cutlet_None  \
count                   30000.000000   30000.000000  30000.000000   
mean                        0.000300       0.001667      0.978333   
std                         0.017318       0.040791      0.145595   
min                         0.000000       0.000000      0.000000   
25%                         0.000000       0.000000      1.000000   
50%                         0.000000       0.000000      1.000000   
75%                         0.000000       0.000000      1.000000   
max                         1.000000       1.000000      1.000000   

       cutlet_Pointed  cutlet_Slightly Large  cutlet_Small  cutlet_Very Small  
count    30000.000000           30000.000000  30000.000000       30000.000000  
mean         0.000633               0.000300      0.005800           0.013233  
std          0.025159               0.017318      0.075938           0.114274  
min          0.000000               0.000000      0.000000           0.000000  
25%          0.000000               0.000000      0.000000           0.000000  
50%          0.000000               0.000000      0.000000           0.000000  
75%          0.000000               0.000000      0.000000           0.000000  
max          1.000000               1.000000      1.000000           1.000000  

[8 rows x 125 columns]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_rescaled, y, test_size=0.3, random_state=1)

In [ ]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

knn = KNeighborsRegressor(n_neighbors=5, metric='euclidean')
knn.fit(X_train, y_train)

# Make predictions for regression
y_pred = knn.predict(X_test)

# Calculate regression metrics
mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False)
r2 = r2_score(y_test, y_pred)
max_err = max_error(y_test, y_pred)


# Print the regression metrics
print("R-squared (R2) Score:", r2)
print("Mean Absolute Error (MAE):", mae)
print("Root Mean Squared Error (RMSE):", rmse)
print("Maximum Error:", max_err)

R-squared (R2) Score: 0.004058979869106083
Mean Absolute Error (MAE): 4167.408288888888
Root Mean Squared Error (RMSE): 20957.020745135083
Maximum Error: 1034638.0


In [ ]:
from sklearn.model_selection import cross_val_score

max_K = 50
cv_scores = []

for K in range(1, max_K):
    knn = KNeighborsRegressor(n_neighbors=K)
    scores = cross_val_score(knn, X_train, y_train, cv=5, scoring="r2")
    cv_scores.append(scores.mean())

print('Optimal k: ', cv_scores.index(max(cv_scores))+1)

Optimal k:  3


In [ ]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

knn = KNeighborsRegressor(n_neighbors=3, metric='euclidean')
knn.fit(X_train, y_train)

# Make predictions for regression
y_pred = knn.predict(X_test)

# Calculate regression metrics
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False)
max_err = max_error(y_test, y_pred)


# Print the regression metrics
print("R-squared (R2) Score:", r2)
print("Mean Absolute Error (MAE):", mae)
print("Root Mean Squared Error (RMSE):", rmse)
print("Maximum Error:", max_err)

R-squared (R2) Score: -0.06482424804965947
Mean Absolute Error (MAE): 4172.4788148148145
Root Mean Squared Error (RMSE): 21669.640147541693
Maximum Error: 1033488.3333333334


In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

dt_model = DecisionTreeRegressor(min_samples_leaf=5, max_depth=3)
dt_model.fit(X_train, y_train)

# Make predictions for regression
y_pred = dt_model.predict(X_test)

# Calculate regression metrics
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False)
max_err = max_error(y_test, y_pred)

# Print the regression metrics
print("R-squared (R2) Score:", r2)
print("Mean Absolute Error (MAE):", mae)
print("Root Mean Squared Error (RMSE):", rmse)
print("Maximum Error:", max_err)

R-squared (R2) Score: 0.5095228210133348
Mean Absolute Error (MAE): 4116.948697903207
Root Mean Squared Error (RMSE): 14706.933742364614
Maximum Error: 589203.5


In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

rf_model = RandomForestRegressor(max_depth=5, random_state=0)
rf_model.fit(X_train, y_train)

# Make predictions for regression
y_pred_rf = rf_model.predict(X_test)

# Calculate regression metrics
mae_rf = mean_absolute_error(y_test, y_pred_rf)
rmse_rf = mean_squared_error(y_test, y_pred_rf, squared=False)
r2_rf = r2_score(y_test, y_pred_rf)
max_err_rf = max_error(y_test, y_pred_rf)

# Print the regression metrics
print("R-squared (R2) Score for Random Forest:", r2_rf)
print("Mean Absolute Error (MAE) for Random Forest:", mae_rf)
print("Root Mean Squared Error (RMSE) for Random Forest:", rmse_rf)
print("Maximum Error for Random Forest:", max_err_rf)

R-squared (R2) Score for Random Forest: 0.7086399037470974
Mean Absolute Error (MAE) for Random Forest: 1998.3588278780596
Root Mean Squared Error (RMSE) for Random Forest: 11335.162987012161
Maximum Error for Random Forest: 632537.5366387002


In [ ]:
from sklearn.ensemble import AdaBoostRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

bt_model = AdaBoostRegressor(n_estimators=100)
bt_model.fit(X_train, y_train)

# Make predictions for regression
y_pred_bt = bt_model.predict(X_test)

# Calculate regression metrics
mae_bt = mean_absolute_error(y_test, y_pred_bt)
rmse_bt = mean_squared_error(y_test, y_pred_bt, squared=False)
r2_bt = r2_score(y_test, y_pred_bt)
max_err_bt = max_error(y_test, y_pred_bt)

# Print the regression metrics
print("R-squared (R2) Score for Boosted Tree:", r2_bt)
print("Mean Absolute Error (MAE) for Boosted Tree:", mae_bt)
print("Root Mean Squared Error (RMSE) for Boosted Tree:", rmse_bt)
print("Maximum Error for Boosted Tree:", max_err_bt)

R-squared (R2) Score for Boosted Tree: 0.5862170842829424
Mean Absolute Error (MAE) for Boosted Tree: 5200.6880281156555
Root Mean Squared Error (RMSE) for Boosted Tree: 13508.247554225565
Maximum Error for Boosted Tree: 638274.0272727273


In [ ]:
import xgboost as xgb
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Create and train an XGBoost regression model
xgb_model = xgb.XGBRegressor()
xgb_model.fit(X_train, y_train)

# Make predictions for regression
y_pred_xgb = xgb_model.predict(X_test)

# Calculate regression metrics
mae_xgb = mean_absolute_error(y_test, y_pred_xgb)
rmse_xgb = mean_squared_error(y_test, y_pred_xgb, squared=False)
r2_xgb = r2_score(y_test, y_pred_xgb)
max_err_xgb = max_error(y_test, y_pred_xgb)

# Print the regression metrics
print("R-squared (R2) Score for XGBoost:", r2_xgb)
print("Mean Absolute Error (MAE) for XGBoost:", mae_xgb)
print("Root Mean Squared Error (RMSE) for XGBoost:", rmse_xgb)
print("Maximum Error for XGBoost:", max_err_xgb)

R-squared (R2) Score for XGBoost: 0.7886127567830344
Mean Absolute Error (MAE) for XGBoost: 1022.0985562260946
Root Mean Squared Error (RMSE) for XGBoost: 9654.996500662885
Maximum Error for XGBoost: 495296.5625



- XGBoost excels based on the provided metrics:
  - It achieves a high R-squared (R2) score (0.7886), indicating strong predictive power.
  - It has the lowest Mean Absolute Error (MAE - 1022.10), signifying high accuracy.
  - With the lowest Root Mean Squared Error (RMSE - 9655.00), it demonstrates reliability.
  - The low Maximum Error (495296.56) indicates minimal extreme prediction errors.
- Collectively, these metrics highlight XGBoost's effectiveness in making accurate predictions for the regression task.